### DISCUSSION TITLE
[How to] Debounced atom?

### DISCUSSION THREAD

**Discussion by valtism** [pmndrs/jotai#553] [Q&A] [answered]
Hi there! 

I've been switching a personal project over to Jōtai, and I'm loving it so far.

The issue I've been stuck on is how to debounce within atoms. In my use-case, I am hitting a search API and would like to debounce the input so we don't slam the API on every keystroke. 

I can do this easily enough in react, although it causes extra renders when the debounce hook return, and also feels a bit wrong. I would like to have this debounce happen _within_ the atoms, if that is possible.

Here is a sandbox of the basics of what I am trying to do, without any debouncing: https://codesandbox.io/s/confident-diffie-308h2?file=/src/App.js

I am not sure how I would achieve this. If I need an intermediate atom to handle the debounced query or what. Any help would be greatly appreciated :)

**Comment by Aslemammad** [comment, marked as answer]
Hello @valtism! Thank you for your support & love to jotai!
Here's the solution:
```ts
import debounce from "debounce-async";

const inputAtom = atom("");

const debouncedFetch = debounce(async (query) => {
  const response = await fetch(`https://swapi.dev/api/people/?search=${query}`);
  return await response.json();
}, 500);

const fetchAtom = atom((get) => debouncedFetch(get(inputAtom)));
/*
   You can do this too. Suggested by daishi.

   const query = get(inputAtom);
   if (!query) return;
   return debouncedFetch(query);
*/
```
[codesandbox](https://codesandbox.io/s/brave-shirley-9wwc2?file=/src/App.js)

**Reply by dai-shi** [reply]
Yeah, this looks good, if it's okay to show loading immediately.

Slightly updated not to show loading initially.
https://codesandbox.io/s/bold-cherry-yqmh4

**Reply by Aslemammad** [reply]
Yea, that's a brilliant one.

**Reply by valtism** [reply]
Thank you so much guys! I was hesitant in the answer at first because I didn't understand the intricacies of the debounce package. I was, however, able to write a debounce myself that looks like this:

```ts
function debounce<T extends unknown[], U>(
  callback: (...args: T) => PromiseLike<U> | U,
  wait: number
) {
  let timer: number;

  return (...args: T): Promise<U> => {
    clearTimeout(timer);
    return new Promise((resolve) => {
      timer = setTimeout(() => resolve(callback(...args)), wait);
    });
  };
}
```

And am able to use it the way you proposed. Works great!

**Comment by austinwoon** [comment]
@dai-shi Would there be any value add in having a new utility hook `useDebouncedAtom`? Debouncing is a pretty common use-case, and we needed it in various part of our project. I created this utility function and thought to share in case you felt it might be useful for jotai users.

```tsx
export const useDebouncedAtom = <T,>(
  initialValue: T,
  atomConfig: PrimitiveAtom<T>,
  delayMilliseconds = 500,
  shouldDebounceOnReset = false,
  onDebouncing: () => void,
  onDebounceResolve: () => void,
): [T, Dispatch<SetStateAction<T>>] => {
  const [value, setValue] = useState<T>(initialValue)
  const setGlobalValue = useSetAtom(atomConfig)
  const resetConditionMet = value === initialValue

  useEffect(() => {
    // if reset condition is met, do not debounce
    if (resetConditionMet && !shouldDebounceOnReset) {
      setGlobalValue(value)
      onDebounceResolve()
      return
    }

    onDebouncing()

    const timeoutId = setTimeout(() => {
      setGlobalValue(value)
      onDebounceResolve()
    }, delayMilliseconds)

    return () => clearTimeout(timeoutId)
  }, [
    onDebounceResolve,
    onDebouncing,
    shouldDebounceOnReset,
    delayMilliseconds,
    resetConditionMet,
    setGlobalValue,
    value,
  ])

  return [value, setValue]
}
```

**Simple usage**
```tsx
const nameAtom = atom('')

const MyComponent = () => {
  const [value, setValue] = useDebouncedAtom<string>('', nameAtom)

  return <input onChange={(e) => setValue(e.currentTarget.value)}></input>
}
```

**Passing in custom callbacks so that we know when state is debouncing**
```tsx
const nameAtom = atom('')
const debouncingAtom = atom(false)

const MyComponent = () => {
  const [debouncing, setDebouncing] = useAtom(debouncingAtom)

  const onDebouncing = useCallback(() => setDebouncing(true), [setDebouncing])
  const onDebounceResolve = useCallback(
    () => setDebouncing(false),
    [setDebouncing],
  )
  const [value, setValue] = useDebouncedAtom<string>(
    '',
    nameAtom,
    500,
    false,
    onDebouncing,
    onDebounceResolve,
  )

  return (
    <>
      {debouncing ?? 'Loading'}
      {value}
      <input onChange={(e) => setValue(e.currentTarget.value)} />
    </>
  )
}
```

**Reply by dai-shi** [reply]
A few notes:
- We generally prefer atom utils instead of hook utils.
- Jotai repo is getting too big so I'd prefer [jotai-labs](https://github.com/jotai-labs) for new utils.
- I'm not very sure about the goal but there would be a better implementation.
- Especially, the use of `useEffect` doesn't seem nice with recent practice. cf. https://twitter.com/dai_shi/status/1554467083089747968

**Reply by austinwoon** [reply]
Thanks for the insights! Took your points in consideration + relooked at the existing utils and came up with the code snippet below.

Could i ask the following?:
- Is it safe to use `SECRET_INTERNAL_getScopeContext`? Im trying to follow the behavior of current atom utils and all of them use this exported function, however it was noted that this was for internal use only. 
- A follow up to the previous question, should I just use the standard useAtom hook instead of using this method to set state? I wasnt sure of the implications of using either method but noted most utils had set state with the implementation below.
- My vite configs can't seem to find exports from `core/store`. Is it intentional that constants (such as WRITE_ATOM) do not get exported?

```tsx
import { MutableRefObject, useContext, useRef } from 'react'
import { useCallback } from 'react'
import { WritableAtom } from 'jotai'
// it was explicitly said that SECRET_INTERNAL_getScopeContext is not recommended for app usage, to check with Daishi
import { SECRET_INTERNAL_getScopeContext as getScopeContext } from 'jotai'
import { Scope } from 'jotai/core/atom'

// this const is inferred from  `WRITE_ATOM` exported in 'jotai/core/store'.
// We get a vite error when trying to import directly so doing this make shfit
const WRITE_ATOM_KEY = 'w'

export function useAtomWithDebounce<Value, Update>(
  anAtom: WritableAtom<Value, Update>,
  delayMilliseconds: number,
  scope?: Scope,
) {
  const timeout: MutableRefObject<null | NodeJS.Timeout> = useRef(null)
  const ScopeContext = getScopeContext(scope)
  const store = useContext(ScopeContext).s
  const setAtom = useCallback(
    (update: Update) => {
      if (timeout.current) {
        clearTimeout(timeout.current)
      }

      timeout.current = setTimeout(() => {
        store[WRITE_ATOM_KEY](anAtom, update)
      }, delayMilliseconds)
    },
    [timeout, delayMilliseconds, store, anAtom],
  )
  return setAtom
}
```

**Reply by austinwoon** [reply]
The idea for this utility is simple - delay updates to the atom for duration of `delayMilliseconds`.

**Reply by austinwoon** [reply]
Ignore the above! Went with the simplest approach by defining my own atom creator. 

```tsx
export function atomWithDebounce<Value>(
  initialValue: Value,
  delayMilliseconds = 500,
  shouldDebounceOnReset = false,
) {
  let timeout: NodeJS.Timeout

  const valueAtom = atom(initialValue)

  const debouncedValueAtom = atom(
    initialValue,
    (get, set, update: SetStateAction<Value>) => {
      clearTimeout(timeout)
      const prevValue = get(valueAtom)
      const nextValue =
        typeof update === 'function'
          ? (update as (prev: Value) => Value)(prevValue)
          : update

      set(valueAtom, nextValue)

      if (!shouldDebounceOnReset && nextValue === initialValue) {
        set(debouncedValueAtom, nextValue)
        return
      }

      set(valueAtom, nextValue)

      timeout = setTimeout(() => {
        set(debouncedValueAtom, nextValue)
      }, delayMilliseconds)
    },
  )

  const isDebouncingAtom = atom(
    (get) => get(valueAtom) !== get(debouncedValueAtom),
  )

  return {
    valueAtom,
    isDebouncingAtom,
    debouncedValueAtom,
  }
}
```

**Reply by dai-shi** [reply]
`atomWithDebounce` looks cool. This is exactly what I mean by "atom util".
There can be some variants (or simpler one), so it's probably worth adding as an advanced recipe in docs.
One problem is that `timeout` is defined outside atom, which means it's module state. This causes an unexpected behavior with multiple Provider use case.
I will be adding a new feature for abortability #1091, which may or may not help.

---

> Is it safe to use SECRET_INTERNAL_getScopeContext?

It's an internal function. Basically, we should better avoid it to create a util function.

> should I just use the standard useAtom hook instead of using this method to set state?

Yes. We should use public APIs.

> Is it intentional that constants (such as WRITE_ATOM) do not get exported?

Yes. Those are implementation details and not exported intentionally.

**Reply by austinwoon** [reply]
Cool! I will proceed to open a PR for this in `advanced-recipes` docs. Thanks again for guidance on what the proper approach should be with `jotai`. It'll help in our current project too.

Will take a look at how i can update `timeout` to not be in module scope.

**Reply by austinwoon** [reply]
> One problem is that timeout is defined outside atom, which means it's module state. This causes an unexpected behavior with multiple Provider use case.

Could I check what unexpected behavior could occur? Did not manage to find any sources online that would help me understand the issue.

**Reply by dai-shi** [reply]
It's not easy to reproduce the expected behavior because it depends on timing. Do you understand what the multiple Provider use case is? Here's a simple counter example: https://twitter.com/dai_shi/status/1447892237753278466

**Reply by austinwoon** [reply]
yep! we use multiple providers in our projects so i understand it. 

I interpreted the your phrasing as there could be possible collisions on `timeoutId` in different providers, but thought that shouldn't occur. Im not sure I understand what the issue is with multiple providers in this case.

**Reply by dai-shi** [reply]
> there could be possible collisions on `timeoutId` in different providers, but thought that shouldn't occur.

Yes. It should occur.

**Comment by AjaxSolutions** [comment]
I asked in the PR, how do I cancel debounce cycle?

See [useDebouncedCallback](https://github.com/xnimorz/use-debounce#advanced-usage)

**Reply by austinwoon** [reply]
I can probably store timeoutId as an atom then expose an atom setter to clear the timeout whenever it gets called.

**Reply by austinwoon** [reply]
Did a sample implementation in PR I opened, but also replying here in case it might be useful for others.

https://codesandbox.io/s/elegant-mcclintock-nt28hi
